### PR TITLE
Add an optional pyTorch `device` parameter to isn() functions

### PR DESCRIPTION
**Pull request by GiadaLalli**
no dear clue if this will ever work; I also opened an issue some time ago, dunno if it's still there, but I already has listed in there some problems I was facing with the library, e.g. we should be able to the isn-computation functions whatever metric we are pleased with, but that was not working by the time I tried it; also the metrics need to be changed somehow - how I wanna keep all of them tbh, I just want to have as "fixed" metrics the fastest ones but then I wanna keep also the others to be able to choose (this makes more sense in my brain than written all out here). Once all this is proved to be working (and hopefully very very very soon), I can move on to the benchmark (FINALLY) and then to finish that awfully stupid paper and shall my life be compelte? Who knows.

### CODE REVIEW

File: isn_tractor/metrics.py [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+import torch as t
+from torchmetrics import SpearmanCorrCoef, PearsonCorrCoef, R2Score
+
+
+def pearson_metric_t(first: t.Tensor, second: t.Tensor):
+    "TODO:"
+    if (first.dim(), second.dim()) == (1, 1):
+        return t.corrcoef(t.tensor(first, second))[
+            0, 1
+        ]  # TODO: I have no idea what you're doing here
```
Comment (MaybeJustJames) [inline isn_tractor/metrics.py:16]: Was this intended to be a `t.cat`?

File: isn_tractor/metrics.py
Comment (MaybeJustJames) [inline isn_tractor/metrics.py:1]: I don't really think this file belongs in the project. These could be user specified functions. If you've added them only for benchmarking then IMHO these functions belong in your (separate) benchmarking code.

Comment (MaybeJustJames) [reply isn_tractor/metrics.py:1]: If you want to keep them in the project they need unit tests.

File: isn_tractor/ibisn.py [hunk 2]
```diff
@@ -270,8 +270,8 @@ def isn_edge_implementation(first: t.Tensor, second: t.Tensor):
         result = []
 
         for indx in range(first.shape[0]):
-            loo_1 = t.tensor(np.delete(first, indx, axis=0))
-            loo_2 = t.tensor(np.delete(second, indx, axis=0))
+            loo_1 = t.cat((first[:indx], first[indx + 1 :]))
+            loo_2 = t.cat((second[:indx], second[indx + 1 :]))
```
Comment (MaybeJustJames) [inline isn_tractor/ibisn.py:274]: This removes the warning from pyTorch.

**Review by MaybeJustJames — CHANGES_REQUESTED** [review]

File: isn_tractor/metrics.py
Comment (MaybeJustJames) [reply isn_tractor/metrics.py:1]: Actually. Now that I think about it.... it would make sense to include a "benchmark" directory. So your benchmarking code could reasonably be included in this repository. In which case, I believe this file belongs in a `benchmark/` folder.

Comment (GiadaLalli) [reply isn_tractor/metrics.py:1]: Actually I need to switch the metrics with Torch-implemented or TorchMetrics, to make the code run faster and to give a sense to the CUDA parameter

Comment (MaybeJustJames) [reply isn_tractor/metrics.py:1]: I see. Can we, instead, benchmark? Then use whichever are fastest?

File: isn_tractor/ibisn.py [hunk 3]
```diff
@@ -225,28 +222,15 @@ def map_interaction(
 
 # ## Metrics for unmapped discrete data
 
-
 def __pearson_metric(first, second):
```
Comment (GiadaLalli) [inline isn_tractor/ibisn.py:225]: reminder for myself:

def __pearson_metric(first: t.Tensor, second: t.Tensor) -> t.Tensor:
    if first.dim() == 1 and second.dim() == 1:
        combined = t.stack([first, second], dim=1)
    else:
        combined = t.cat([first, second], axis=1) 
    return t.corrcoef(combined.T)[: first.shape[1] - 1, first.shape[1] :]

File: isn_tractor/ibisn.py [hunk 4]
```diff
@@ -398,10 +398,32 @@ def dense_isn(data: pd.DataFrame, metric: Metric):
     dense.iloc[:, 0] = np.repeat(net.columns.values, net.columns.size)
     dense.iloc[:, 1] = np.tile(net.columns.values, data.shape[0])
 
+    if device:
+        device = t.device("cuda" if t.cuda.is_available() else "cpu")
+    else:
+        device = t.device("cpu")
+
```
Comment (MaybeJustJames) [inline isn_tractor/ibisn.py:405]:
```suggestion
```

File: isn_tractor/ibisn.py [hunk 5]
```diff
@@ -398,10 +398,32 @@ def dense_isn(data: pd.DataFrame, metric: Metric):
...
+                .numpy()
+                .flatten()
+            )
+        else:
+            values = metric_fn(
+                pd.DataFrame(np.delete(data.T.to_numpy(), i, 0))
+            ).values.flatten()
+        dense.iloc[:, i + 2] = num_samples * (agg - values) + values
+
+    for i in range(num_samples):
```
Comment (MaybeJustJames) [inline isn_tractor/ibisn.py:420]:
```suggestion
```

File: isn_tractor/ibisn.py [hunk 6]
```diff
@@ -284,7 +287,12 @@ def __make_array(*xs):
     return np.array(xs, dtype=object)
 
 
-def __make_edge_fn(data, metric_fn: MetricFn, pool_fn: PoolingFn):
+def __make_edge_fn(
+    data,
+    metric_fn: MetricFn,
+    pool_fn: PoolingFn,
```
Comment (GiadaLalli) [inline isn_tractor/ibisn.py:293]: I have the feeling that the __pooling function - which I can't find anymore but probably it's just me - is what is making the tests fail, but may be just a stupid hypothesis.

Comment (MaybeJustJames) [reply isn_tractor/ibisn.py:293]: :) it's right there where you've highlighted: `pool_fn`.

Comment (GiadaLalli) [reply isn_tractor/ibisn.py:293]: I didn't express myself correctly; what I meant is that it's not defined before; it's like we are calling the pooling function when there's no pooling function defined. Dunno if I was clearer now - and dunno if that's even the actual problem tho.

Comment (MaybeJustJames) [reply isn_tractor/ibisn.py:293]: Ah I see. https://github.com/GiadaLalli/ISN-tractor/blob/cuda_impl/isn_tractor/ibisn.py#L358

**Comment by GiadaLalli** [comment]
`def Spearman(first, second):
    if first.ndim == 1 and second.ndim == 1:
        data = t.stack((first, second), dim=1)
        for i in range(data.shape[1]):
            u, inv, counts = t.unique(data[:,i], return_inverse=True, return_counts=True)
            csum = t.zeros_like(counts)
            csum[1:] = counts[:-1].cumsum(dim=-1)
            data[:,i] = csum[inv]
        corr = t.corrcoef(data.T)[0, 1]
    else:
        data = t.cat((first, second), dim=1)
        for i in range(data.shape[1]):
            u, inv, counts = t.unique(data[:,i], return_inverse=True, return_counts=True)
            csum = t.zeros_like(counts)
            csum[1:] = counts[:-1].cumsum(dim=-1)
            data[:,i] = csum[inv]
        corr = t.corrcoef(data.T)[:first.shape[1], first.shape[1]:]
    return(corr)`

**Comment by GiadaLalli** [comment]
`This version of _def spearman_corr(first, second):_ gives the same results as the Scipy version:

    if first.ndim == 1 and second.ndim == 1:
        data = t.stack((first, second), dim=1)
        for i in range(data.shape[1]):
            u, inv, counts = t.unique(data[:,i], return_inverse=True, return_counts=True)
            csum = t.zeros_like(counts)
            csum[1:] = counts[:-1].cumsum(dim=-1)
            data[:,i] = csum[inv]
        corr = t.corrcoef(data.T)[0, 1]
    else:
        data = t.cat((first, second), dim=1)
        for i in range(data.shape[1]):
            u, inv, counts = t.unique(data[:,i], return_inverse=True, return_counts=True)
            csum = t.zeros_like(counts)
            csum[1:] = counts[:-1].cumsum(dim=-1)
            data[:,i] = csum[inv]
        corr = t.corrcoef(data.T)[:first.shape[1], first.shape[1]:]
    return(corr)`

Here the values used as array/tensor used as inputs:

````
first = [[0.6270, 0.9965, 0.0072, 0.1089], 
[0.2653, 0.3640, 0.8955, 0.4664], 
[0.4866, 0.6505, 0.9105, 0.1444]]

second = [[0.6416, 0.0196, 0.6138, 0.2514, 0.3511], 
[0.9249, 0.4233, 0.7246, 0.9369, 0.3315], 
[0.1313, 0.1769, 0.1377, 0.8924, 0.4274]]

_____________________________________________________

first = tensor([[0.6270, 0.9965, 0.0072, 0.1089],
                [0.2653, 0.3640, 0.8955, 0.4664],
                [0.4866, 0.6505, 0.9105, 0.1444]])
second = tensor([[0.6416, 0.0196, 0.6138, 0.2514, 0.3511],
                [0.9249, 0.4233, 0.7246, 0.9369, 0.3315],
                [0.1313, 0.1769, 0.1377, 0.8924, 0.4274]])

File: pyproject.toml [hunk 7]
```diff
@@ -18,6 +18,7 @@ numpy = "^1.24.1"
 scipy = "^1.10.0"
 scikit-learn = "^1.2.0"
 torch = "^1.13.1"
+torchmetrics = "^0.11.4"
```
Comment (MaybeJustJames) [inline pyproject.toml:21]:
```suggestion
```

**Review by MaybeJustJames — APPROVED** [review]
Good work!
